File: backend/src/main/java/com/locationservice/controller/AuthController.java
package com.locationservice.controller;

import com.locationservice.dto.LoginRequest;
import com.locationservice.dto.LoginResponse;
import com.locationservice.dto.RegisterRequest;
import com.locationservice.dto.RegisterResponse;
import com.locationservice.dto.UserResponse;
import com.locationservice.entity.RefreshToken;
import com.locationservice.entity.User;
import com.locationservice.exception.TokenRefreshException;
import com.locationservice.service.AuthService;
import com.locationservice.service.RefreshTokenService;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final AuthService authService;
    private final RefreshTokenService refreshTokenService;

    public AuthController(AuthService authService, RefreshTokenService refreshTokenService) {
        this.authService = authService;
        this.refreshTokenService = refreshTokenService;
    }

    @PostMapping("/register")
    public ResponseEntity<RegisterResponse> register(@Valid @RequestBody RegisterRequest request) {
        RegisterResponse response = authService.register(request);
        return new ResponseEntity<>(response, HttpStatus.CREATED);
    }

    @PostMapping("/login")
    public ResponseEntity<LoginResponse> login(@Valid @RequestBody LoginRequest request) {
        User user = authService.authenticate(request);
        RefreshToken refreshToken = refreshTokenService.createRefreshToken(user);

        ResponseCookie accessCookie = authService.createAccessTokenCookie(user);
        ResponseCookie refreshCookie = authService.createRefreshTokenCookie(refreshToken);

        LoginResponse response = new LoginResponse(
                "Login successful",
                user.getId(),
                user.getName(),
                user.getEmail(),
                user.getRole());

        return ResponseEntity.ok()
                .header(HttpHeaders.SET_COOKIE, accessCookie.toString())
                .header(HttpHeaders.SET_COOKIE, refreshCookie.toString())
                .body(response);
    }

    @PostMapping("/refresh")
    public ResponseEntity<Map<String, String>> refreshToken(
            @CookieValue(name = "refresh_token", required = false) String refreshTokenFromCookie,
            HttpServletRequest request) {

        String token = refreshTokenFromCookie;
        if ((token == null || token.isBlank()) && request.getCookies() != null) {
            for (Cookie cookie : request.getCookies()) {
                if ("refresh_token".equals(cookie.getName())) {
                    token = cookie.getValue();
                    break;
                }
            }
        }

        if (token == null || token.isBlank()) {

            throw new TokenRefreshException("Refresh token is missing. Please sign in again.");
        }

        RefreshToken refreshToken = refreshTokenService.findByToken(token)
                .orElseThrow(() -> new TokenRefreshException("Refresh token not found. Please sign in again."));

        refreshTokenService.verifyExpiration(refreshToken);
        User user = refreshToken.getUser();

        // Rotate Refresh Token & Issue New Access Token
        RefreshToken newRefreshToken = refreshTokenService.createRefreshToken(user);
        ResponseCookie newAccessCookie = authService.createAccessTokenCookie(user);
        ResponseCookie newRefreshCookie = authService.createRefreshTokenCookie(newRefreshToken);

        return ResponseEntity.ok()
                .header(HttpHeaders.SET_COOKIE, newAccessCookie.toString())
                .header(HttpHeaders.SET_COOKIE, newRefreshCookie.toString())
                .body(Map.of("message", "Token refreshed successfully"));
    }

    @PostMapping("/logout")
    public ResponseEntity<Map<String, String>> logout(
            @CookieValue(name = "refresh_token", required = false) String refreshTokenFromCookie,
            Authentication authentication) {

        if (refreshTokenFromCookie != null && !refreshTokenFromCookie.isBlank()) {
            refreshTokenService.deleteByToken(refreshTokenFromCookie);
        } else if (authentication != null && authentication.getPrincipal() instanceof User user) {
            refreshTokenService.deleteByUser(user);
        }

        ResponseCookie deleteAccessCookie = authService.createLogoutAccessCookie();
        ResponseCookie deleteRefreshCookie = authService.createLogoutRefreshCookie();

        return ResponseEntity.ok()
                .header(HttpHeaders.SET_COOKIE, deleteAccessCookie.toString())
                .header(HttpHeaders.SET_COOKIE, deleteRefreshCookie.toString())
                .body(Map.of("message", "Logged out successfully"));
    }

    @GetMapping("/me")
    public ResponseEntity<UserResponse> getCurrentUser(Authentication authentication) {
        User user = (User) authentication.getPrincipal();
        UserResponse response = new UserResponse(
                user.getId(),
                user.getName(),
                user.getEmail(),
                user.getRole());
        return ResponseEntity.ok(response);
    }
}

File: backend/src/controllers/auth.controller.ts
import { Request, Response } from 'express';
import { AuthService } from '../services/auth.service.js';
import { JwtService } from '../services/jwt.service.js';

export class AuthController {
  public static async register(req: Request, res: Response): Promise<void> {
    try {
      const response = await AuthService.register(req.body);
      res.status(201).json(response);
    } catch (error: any) {
      res.status(400).json({ error: 'Bad Request', message: error.message });
    }
  }

  public static async login(req: Request, res: Response): Promise<void> {
    try {
      const user = await AuthService.authenticate(req.body.email, req.body.password);
      const refreshToken = await AuthService.createRefreshToken(user.id);
      const accessToken = JwtService.generateAccessToken(user.id, user.email, user.role);

      JwtService.setAccessTokenCookie(res, accessToken);
      JwtService.setRefreshTokenCookie(res, refreshToken.token);

      res.status(200).json({
        message: 'Login successful',
        userId: Number(user.id),
        name: user.name,
        email: user.email,
        role: user.role,
      });
    } catch (error: any) {
      res.status(401).json({ error: 'Unauthorized', message: error.message });
    }
  }

  public static async refreshToken(req: Request, res: Response): Promise<void> {
    try {
      let token = req.cookies?.refresh_token;

      if (!token) {
        res.status(401).json({ error: 'Unauthorized', message: 'Refresh token is missing. Please sign in again.' });
        return;
      }

      const refreshToken = await AuthService.verifyRefreshToken(token);
      const newRefreshToken = await AuthService.createRefreshToken(refreshToken.user.id);
      const newAccessToken = JwtService.generateAccessToken(refreshToken.user.id, refreshToken.user.email, refreshToken.user.role);

      JwtService.setAccessTokenCookie(res, newAccessToken);
      JwtService.setRefreshTokenCookie(res, newRefreshToken.token);

      res.status(200).json({ message: 'Token refreshed successfully' });
    } catch (error: any) {
      res.status(401).json({ error: 'Unauthorized', message: error.message });
    }
  }

  public static async logout(req: Request, res: Response): Promise<void> {
    try {
      const refreshToken = req.cookies?.refresh_token;
      await AuthService.deleteRefreshToken(refreshToken, req.user?.id);

      JwtService.clearAuthCookies(res);
      res.status(200).json({ message: 'Logged out successfully' });
    } catch (error: any) {
      res.status(500).json({ error: 'Internal Server Error', message: error.message });
    }
  }

  public static async getCurrentUser(req: Request, res: Response): Promise<void> {
    if (!req.user) {
      res.status(401).json({ error: 'Unauthorized' });
      return;
    }

    res.status(200).json({
      userId: Number(req.user.id),
      name: req.user.name,
      email: req.user.email,
      role: req.user.role,
    });
  }
}
